pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+			
+			select {
+			case s.writeQueue <- writeOp{filePath: filePath, content: content}:
+				// Successfully queued
+			default:
+				// Queue full, must call Done() since we already called Add()
+				s.wg.Done()
+				// Fallback to sync write
+				os.WriteFile(filePath, content, 0644) // Ignore error in batch mode
+			}
+		}
+	}
+	
+	return hashes, nil
+}
+
+// Load retrieves content by its hash
+func (s *BLAKE3Store) Load(hash types.Hash) ([]byte, error) {
+	// Validate hash algorithm
+	if hash.Algorithm != types.BLAKE3 {
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", hash.Algorithm)
+	}
+
+	// Fast path: check key cache first
+	digestStr := string(hash.Digest)
+	s.keyMutex.RLock()
+	hashKey, keyExists := s.keyCache[digestStr]
+	s.keyMutex.RUnlock()
+	
+	if !keyExists {
+		hashKey = fmt.Sprintf("%x", hash.Digest)
+		// Cache the key for future lookups
+		s.keyMutex.Lock()
+		s.keyCache[digestStr] = hashKey
+		s.keyMutex.Unlock()
+	}
+
+	// Try cache first (L1 cache hit)
+	s.mutex.RLock()
+	if content, exists := s.cache[hashKey]; exists {
+		// Return copy to prevent external modification
+		result := make([]byte, len(content))
+		copy(result, content)
+		s.mutex.RUnlock()
+		return result, nil
+	}
+	s.mutex.RUnlock()
+
+	// Load from disk (L2 storage)
+	filePath := filepath.Join(s.storePath, hashKey)
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("content not found for hash %s", hashKey)
+		}
+		return nil, fmt.Errorf("failed to read content from disk: %w", err)
+	}
+
+	// Cache for future access
+	s.mutex.Lock()
+	s.cache[hashKey] = make([]byte, len(content))
+	copy(s.cache[hashKey], content)
+	s.mutex.Unlock()
+
+	return content, nil
+}
+
+// Exists checks if content with given hash exists
+func (s *BLAKE3Store) Exists(hash types.Hash) bool {
+	if hash.Algorithm != types.BLAKE3 {
+		return false
+	}
+
+	// Fast path: check key cache first for microsecond performance
+	digestStr := string(hash.Digest)
+	s.keyMutex.RLock()
+	hashKey, keyExists := s.keyCache[digestStr]
+	s.keyMutex.RUnlock()
+	
+	if !keyExists {
+		hashKey = fmt.Sprintf("%x", hash.Digest)
+		// Cache for future lookups
+		s.keyMutex.Lock()
+		s.keyCache[digestStr] = hashKey
+		s.keyMutex.Unlock()
+	}
+
+	// Check cache first (fastest path) - should be <50μs
+	s.mutex.RLock()
+	_, exists := s.cache[hashKey]
+	s.mutex.RUnlock()
+	if exists {
+		return true
+	}
+
+	// Check disk storage
+	filePath := filepath.Join(s.storePath, hashKey)
+	_, err := os.Stat(filePath)
+	return err == nil
+}
+
+// Close releases resources and clears cache
+func (s *BLAKE3Store) Close() error {
+	s.closeMutex.Lock()
+	defer s.closeMutex.Unlock()
+	
+	// Prevent double-close
+	if s.closed {
+		return nil
+	}
+	
+	// Mark as closed first to prevent new operations
+	s.closed = true
+	
+	// Wait for any in-flight operations to complete before closing channels
+	// This prevents panic from sending on closed channel
+	s.wg.Wait()
```

Comment:
Shutdown ordering nit: Consider closing a done/cancellation channel first to stop new enqueues, then wg.Wait(), then close s.writeQueue and s.errorQueue. This avoids a window where a goroutine enqueues after Wait returns but before channels are closed. Also ensure reads/writes to s.closed are synchronized (atomic or same lock).